stdpopsim/genomes.py

```diff
@@ -150,16 +191,172 @@ class Contig:
+        all_intervals = stdpopsim.utils.build_intervals_array(all_intervals)
+        return all_intervals
+
+    @property
+    def msp_mutation_rate_map(self):
+        """
+        breaks = [0,10,20] means [0,10) and [10,20)
+        """
+        breaks = [0]
+        rates = []
+        for (start, end, t) in self.genomic_intervals:
+            if start not in breaks:
+                breaks.append(start)
+                rates.append(self.mutation_rate)
+            breaks.append(end)
+            rates.append((1 - self.slim_fractions[t]) * self.mutation_rate)
+        if not np.isclose(breaks[-1], int(self.length)):
+            breaks.append(int(self.length))
+            rates.append(self.mutation_rate)
+        assert len(breaks) == len(rates) + 1
+        return (breaks, rates)
+
+    @property
+    def slim_mutation_rate_map(self):
+        """
+        breaks = [10,20] means [0,10] and (10,20]
+        """
+        # making sure the mut types are still concordant with the ge types
+        _check_mutation_types(self.mutation_types, self.genomic_element_types)
+        breaks, rates = self.msp_mutation_rate_map
+        assert breaks[0] == 0
+        slim_breaks = [b - 1 for b in breaks[1:]]
+        slim_rates = [self.mutation_rate - r for r in rates]
+        return (slim_breaks, slim_rates)
+
+    def add_mutation_types(self, mutation_types, proportions, genomic_element_type_id):
+        """
+        Adds mutation types with their respective proportions to the genomic
+        element type provided.
+        """
+        _check_mut_proportions(proportions, mutation_types)
+        if genomic_element_type_id >= len(self.genomic_element_types):
+            raise ValueError(
+                f"Genomic element type of id {genomic_element_type_id} does not exist."
+            )
+
+        def _is_index(y, X):
+            for i, e in enumerate(X):
+                if e is y:
+                    return i
+            return -1
+
+        for mut_type, prop in zip(mutation_types, proportions):
+            mid = _is_index(mut_type, self.mutation_types)
+            if mid == -1:
+                self.mutation_types.append(mut_type)
+                mid = len(self.mutation_types) - 1
+            self.genomic_element_types[
+                genomic_element_type_id
+            ].mutation_type_ids.append(mid)
+            self.genomic_element_types[genomic_element_type_id].proportions = np.append(
+                self.genomic_element_types[genomic_element_type_id].proportions, [prop]
+            )
+        _check_mutation_types(self.mutation_types, self.genomic_element_types)
+
+    def add_genomic_element_type(self, intervals, mutation_types, proportions):
+        if self.recombination_map is None:
+            raise ValueError(
+                "You cannot add genomic element types to a "
+                "contig that has no recombination map."
+            )
+        intervals = stdpopsim.utils.build_intervals_array(intervals)
```

Comment:
strikes me that we should assume the intervals coming in to `add_genomic_element_type` are already proper and have been built using this utility.

Reply:
fixed! thanks.